Lib/pathlib.py

```diff
@@ -1287,45 +1287,49 @@ def expanduser(self):
     def walk(self, top_down=True, on_error=None, follow_symlinks=False):
         """Walk the directory tree from this directory, similar to os.walk()."""
         sys.audit("pathlib.Path.walk", self, on_error, follow_symlinks)
-        return self._walk(top_down, on_error, follow_symlinks)
-
-    def _walk(self, top_down, on_error, follow_symlinks):
-        # We may not have read permission for self, in which case we can't
-        # get a list of the files the directory contains. os.walk
-        # always suppressed the exception then, rather than blow up for a
-        # minor reason when (say) a thousand readable directories are still
-        # left to visit. That logic is copied here.
-        try:
-            scandir_it = self._scandir()
-        except OSError as error:
-            if on_error is not None:
-                on_error(error)
-            return
-
-        with scandir_it:
-            dirnames = []
-            filenames = []
-            for entry in scandir_it:
-                try:
-                    is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
-                except OSError:
-                    # Carried over from os.path.isdir().
-                    is_dir = False
-
-                if is_dir:
-                    dirnames.append(entry.name)
-                else:
-                    filenames.append(entry.name)
-
-        if top_down:
-            yield self, dirnames, filenames
-
-        for dirname in dirnames:
-            dirpath = self._make_child_relpath(dirname)
-            yield from dirpath._walk(top_down, on_error, follow_symlinks)
+        stack = [(False, self)]
+
+        while stack:
+            must_yield, top = stack.pop()
+            if must_yield:
+                yield top
+                continue
```

Comment:
```suggestion
            is_path, arg = stack.pop()
            if is_path:
                path = arg
            else:
                yield arg
                continue
```

Using the variable `top` to mean 'top of the walk stack' when we already have have a function argument named `top_down` referring to the 'top of the directory tree' could be misleading. 

Perhaps `arg` would be better, and swap the meaning of the boolean? Not too sure about this.

Reply:
I am a bit confused. top_down is not the top of the directory tree. It is the boolean flag that indicates whether we start from top and go down or not.

Or did I misunderstand what you meant?

Reply:
Sorry for not being clear. I mean that the meaning _of 'top'_ in `top_down` is 'the top of the directory tree'.

Reply:
Fixed!